MessagePort is unexpectedly GC'ed after activity absence

cdumez opened the pull request:
#### a4790ba92ba790c7c332319b48b972b51053b765
<pre>
MessagePort is unexpectedly GC&apos;ed after activity absence
<a href="https://bugs.webkit.org/show_bug.cgi?id=193184">https://bugs.webkit.org/show_bug.cgi?id=193184</a>
rdar://54095480

Reviewed by Ryosuke Niwa and Geoffrey Garen.

Per the HTML specification [1], we should keep the MessagePort alive as long
as it is entangled. This makes sense since the port could receive messages as
long as it is entangled. However, we were missing this check inside
MessagePort::virtualHasPendingActivity().

We do still allow collection when entangled if there is no message event handler,
since it wouldn&apos;t be observable then. Even if a new message would arrive, there
would be no-one on JS side to notify.

[1] <a href="https://html.spec.whatwg.org/multipage/web-messaging.html#ports-and-garbage-collection">https://html.spec.whatwg.org/multipage/web-messaging.html#ports-and-garbage-collection</a>

* LayoutTests/http/tests/messaging/messageport-gc-after-xhr-expected.txt: Added.
* LayoutTests/http/tests/messaging/messageport-gc-after-xhr.html: Added.
* LayoutTests/http/tests/messaging/resources/messageport-pong-after-xhr.html: Added.
Add test coverage. The test is based on the one provided by Yoshiaki Jitsukawa from
Sony.

* Source/WebCore/dom/MessagePort.cpp:
(WebCore::MessagePort::virtualHasPendingActivity const):

Canonical link: <a href="https://commits.webkit.org/256342@main">https://commits.webkit.org/256342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44f591e52dc0a42d4e40110d626ed4fff8e6fc5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105045 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165313 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4757 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33465 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100911 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3476 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82075 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30550 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73387 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20137 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4384 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40928 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42917 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39383 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (approved):
r=me

geoffreygaren commented on Source/WebCore/dom/MessagePort.cpp:
```diff
@@ -350,14 +350,17 @@ bool MessagePort::virtualHasPendingActivity() const
     if (!m_hasMessageEventListener)
         return false;
 
+    if (m_entangled)
```
I think it would be slightly clearer to check isEntangled() here and drop the check for m_isDetached above.

cdumez replied:
But it wouldn't be equivalent / right as far as I can tell.

The current code:
- Return false if detached.
- Return true if entangled.
- Then returns true/false based on other flags

With your proposal:
- Return true if entangled and not detached
- Then returns true/false based on other flags

Because we would no longer early false when m_detached, it would be possible (at least in theory) for this function to return true while m_detached=true, based on the other flags. I believe this would be wrong.

aproskuryakov commented on Source/WebCore/dom/MessagePort.cpp:
```diff
@@ -350,14 +350,17 @@ bool MessagePort::virtualHasPendingActivity() const
+    if (m_entangled)
+        return true;
+
+    // If this object has been idle since the remote port declared itself elgibile for GC, we can GC.
```
typo: elgible.

cdumez replied:
Fixed, thanks.

geoffreygaren commented on Source/WebCore/dom/MessagePort.cpp:
```diff
@@ -350,14 +350,17 @@ bool MessagePort::virtualHasPendingActivity() const
+        return true;
+
+    // If this object has been idle since the remote port declared itself elgibile for GC, we can GC.
+    if (!m_hasHadLocalActivitySinceLastCheck && m_isRemoteEligibleForGC)
+        return false;
```
I'm pretty sure that the logic of m_hasHadLocalActivitySinceLastCheck and m_mightBeEligibleForGC is incorrect. It's not mentioned in the spec. And it doesn't make intuitive sense.

m_isRemoteEligibleForGC, isEntangled(), and m_hasMessageEventListener should cover everything. Maybe m_hasHadLocalActivitySinceLastCheck and m_mightBeEligibleForGC can be removed.

geoffreygaren replied:
Side note: I would rename m_isRemoteEligibleForGC to m_remoteHasActivity.

cdumez replied:
I agree the current code is very complicated and likely not entirely correct. I was planning on following up to attempt to clean this up since this is a little riskier and might be a bigger change too).

geoffreygaren reviewed (approved):
r=me

geoffreygaren reviewed (approved):
r=me

cdumez committed:
MessagePort is unexpectedly GC'ed after activity absence
https://bugs.webkit.org/show_bug.cgi?id=193184
rdar://54095480

Reviewed by Ryosuke Niwa and Geoffrey Garen.

Per the HTML specification [1], we should keep the MessagePort alive as long
as it is entangled. This makes sense since the port could receive messages as
long as it is entangled. However, we were missing this check inside
MessagePort::virtualHasPendingActivity().

We do still allow collection when entangled if there is no message event handler,
since it wouldn't be observable then. Even if a new message would arrive, there
would be no-one on JS side to notify.

[1] https://html.spec.whatwg.org/multipage/web-messaging.html#ports-and-garbage-collection

* LayoutTests/http/tests/messaging/messageport-gc-after-xhr-expected.txt: Added.
* LayoutTests/http/tests/messaging/messageport-gc-after-xhr.html: Added.
* LayoutTests/http/tests/messaging/resources/messageport-pong-after-xhr.html: Added.
Add test coverage. The test is based on the one provided by Yoshiaki Jitsukawa from
Sony.

* Source/WebCore/dom/MessagePort.cpp:
(WebCore::MessagePort::virtualHasPendingActivity const):

Canonical link: https://commits.webkit.org/256342@main
